docs: update docs for shared runtime (DOCS-14256) (#9143)
confluentinc/ksql: docs/reference/server-configuration.md

These configuration parameters control the general behavior of ksqlDB server.
Many parameters can only be set once for the entire server, and must be
specified using the `ksql-server.properties` file.

> You can assign the value of some parameters on a per-persistent query basis

Reviewer: @rodesai does `SET` work on push queries or can you not configure them at all?

Reviewer 2: Good point to generalize to all queries. Also with shared runtime, we changed the definition of "per query", so maybe we need to rephrase a little.
@JimGalasyn , what about something like to make this clear:
...
specified using the `ksql-server.properties` file (for on-prem / standalone). In this case, configuration are applied when the cluster starts.
A subset of these configuration parameters can be applied on a running cluster, either for individual queries (using the SET command or the UI), or for the entire cluster (using ALTER SYSTEM command or the UI). If this is the case, this will be mentioned the list below. Note that only this subset can be edited in Confluent Cloud at this time.

Reviewer: > Note that only this subset can be edited in Confluent Cloud at this time.

@JSeb225 should this say "Note that **this subset can only** be edited in Confluent Cloud at this time"?

by using the `SET` statement. This is indicated in the following parameter
sections with the **Per query** block. For ksqlDB in {{ site.ccloud }}, some 

> parameters can be set only by using the ALTER SYSTEM statement, and this is

Reviewer: ```suggestion
parameters can be set only by using the ALTER SYSTEM statement and apply at the system-level, as
```

Reviewer: Alt suggestion:
```suggestion
parameters can be set only by using the ALTER SYSTEM statement and are applied to all queries running on this cluster, as
```

indicated in the corresponding **Per query** block.

Retrieve the current list of configuration settings by using the

> [SHOW PROPERTIES](/developer-guide/ksqldb-reference/show-properties/) command.

Reviewer: @rodesai why do we have both `SET PROPERTIES` and `LIST PROPERTIES`? Is one of them currently deprecated or going to be phased out?

These configuration parameters control the general behavior of ksqlDB server.
Many parameters can only be set once for the entire server, and must be
specified using the `ksql-server.properties` file (for on-prem / standalone).
In this case, configurations are applied when the cluster starts.

A subset of these configuration parameters can be applied on a running cluster,
either for individual queries (using the SET command or the Confluent Cloud

> Console) or for the entire cluster (using the ALTER SYSTEM command or the

Reviewer: maybe note here that `ALTER SYSTEM` is only available on Confluent Cloud ie not yet applicable on Confluent platform/on-prem installations?

Reviewer: Also, and maybe this is just me being ignorant, but what do you mean by "Confluent Cloud Console" -- I'm guessing that's the UI, is that our official term for it? I think I'm just a bit thrown off by the "console" part of it, that always makes me think 'terminal' or 'command line' interface. But if this is established elsewhere in the docs and would be clear to our users in your opinion then just ignore me 🙂

Author: @ableegoldman Yes, "Confluent Cloud Console" is the official name for the Cloud UI, and I agree, it's confusing. We should go with the industry standard and call it "Confluent Cloud **Portal**".

> ## `ksql.functions.substring.legacy.args`

Reviewer: What is this config? Interesting that it already had a note about being per-query, if so then we would need to include it in the query-level properties list here but I actually can't seem to find any config with this or even a related name in the current codebase. I'm guessing it was deprecated/removed at some point? Should we also remove it from the docs here? cc @MichaelDrogalis who added the docs for it

Author: I removed this puppy.

> **Per query:** yes

Reviewer: @rodesai not sure about this one but I think we actually need to just remove it from the docs entirely. I couldn't even find a reference to it in any of the very old ksql branches

Author: removed

## `ksql.internal.topic.replicas`

> **Per query:** yes

Reviewer: @rodesai should be "no"

## `ksql.persistence.default.format.key`

> **Per query:** yes (for {{ site.ccloud }}, settable with ALTER SYSTEM only)

Reviewer: Actually I think we missed these configs as well, they should be per-query and thus settable via `SET`. You have this backwards here, `ALTER SYSTEM` is for system-level configs which are not per-query. (But as noted in this comment, not all configs which are not per-query can be changed via `ALTER SYSTEM`

Of those not-per-query configs, pretty much those with the prefix `ksql.{streams/consumer/producer}` on this are system properties and settable via `ALTER SYSTEM`, everything else is a system config which can NOT be changed via `ALTER SYSTEM` (I know this is confusing because we just changed this)

Reviewer: @rodesai sounds like this one also does not make sense as a per-query config, agreed?

```suggestion
**Per query:** no
```

## `ksql.persistence.default.format.value`

> **Per query:** yes

Reviewer: @rodesai I just noticed that we seem to list some ksql configs as being per-query already, which we don't currently have in the `queryLevelProperties` list but can be added to it -- ksql configs are not affected by consolidating queries into a single runtime and should continue to be settable via `SET`, where appropriate. However some of these make no sense to list as per-query and should be fixed to say `"No"`

Can you run down this list and sanity check my take on which of these should be added as query-level properties and which of these should be fixed?

Reviewer: Note that we're presumably also missing these from the UI as well and will need to do a quick PR for that as well, once we have a finalized list of the query props. Filed https://confluentinc.atlassian.net/browse/KCI-1753

## `ksql.persistence.wrap.single.values`

> **Per query:** yes

Reviewer: @rodesai to verify

```suggestion
**Per query:** no
```

## `ksql.schema.registry.url`

> **Per query:** yes

Reviewer: @rodesai to verify

```suggestion
**Per query:** no
```

**Per query:** yes (for {{ site.ccloud }}, settable with ALTER SYSTEM only)

The {{ site.sr }} URL path to connect ksqlDB to. To communicate with {{ site.sr }}
over a secure connection, see
[Configure ksqlDB for Secured {{ site.srlong }}](/operate-and-deploy/installation/server-config/security#configure-ksqldb-for-secured-confluent-schema-registry).

## `ksql.service.id`

> **Per query:** yes

Reviewer: @rodesai to verify

```suggestion
**Per query:** no
```

## `ksql.query.pull.enable.standby.reads`

> **Per query:** yes (for {{ site.ccloud }}, settable with ALTER SYSTEM only)

Reviewer: @rodesai given how many of these configs which were labeled as per-query are absolutely not query-level I'm now skeptical that any of these actually do work for individual queries, but this one at least in theory _could_ be applied at the query level. Can you double check whether it actually works that way or not?

```suggestion
**Per query:** no (for {{ site.ccloud }}, settable with ALTER SYSTEM only)
```

If it's _not_ a query-level config, then it at least would be a system property (mutable with `ALTER SYSTEM`) and not a server property, right?

Reviewer: @JimGalasyn  I think this one is also supposed to be a feature flag, in which case it should also be a server property, ie just
```
**Per query:** no 
```

## `ksql.query.pull.max.allowed.offset.lag`

> **Per query:** yes (for {{ site.ccloud }}, settable with ALTER SYSTEM only)

Reviewer: @rodesai same here, I'm skeptical but given that it explicitly mentions using the `SET` command in the docs for this one at least I'm inclined to say it actually is query level

Author: @rodesai, fyi this comment was in a collapsed section, and I almost missed it.

Reviewer: Same here, KQE says intentionally query-level after all

```suggestion
**Per query:** yes
```

## `ksql.query.pull.table.scan.enabled`

> **Per query:** yes (for {{ site.ccloud }}, settable with ALTER SYSTEM only)

Reviewer: @rodesai  ditto above 

```suggestion
**Per query:** yes
```

Reviewer 3: This does not need to be settable per query. It's a feature flag. You arguably shouldn't even be able to set it using `ALTER SYSTEM`.

Reviewer: Makes sense -- feature flags should be server properties, in that case @JimGalasyn this one should just be a "no"

## `ksql.query.pull.interpreter.enabled`

> **Per query:** yes (for {{ site.ccloud }}, settable with ALTER SYSTEM only)

Reviewer: @rodesai also calls out being configurable per-query explicitly in the explanation, so I'm going with "yes"
```suggestion
**Per query:** yes
```

## `ksql.sink.partitions` (Deprecated)

> **Per query:** yes (for {{ site.ccloud }}, settable with ALTER SYSTEM only)

Reviewer: @rodesai to sanity check but I'm guessing it's definitely not. Given that it's deprecated I also wouldn't bother adding it to the list of system properties that can be modified via `ALTER SYSTEM` -- seem reasonable?

```suggestion
**Per query:** no
```

## `ksql.sink.replicas` (Deprecated)

> **Per query:** yes (for {{ site.ccloud }}, settable with ALTER SYSTEM only)

Reviewer: @rodesai to sanity check but I'm guessing it's definitely not. Given that it's deprecated I also wouldn't bother adding it to the list of system properties that can be modified via `ALTER SYSTEM` -- seem reasonable?

```suggestion
**Per query:** no
```
